File: packages/core/src/dna/gene-extractors/state-styling.ts
import { BaseGeneExtractor, type AlleleDefinition, type FileExtractionResult, type DetectedAllele } from './base-extractor.js';

import type { GeneId } from '../types.js';

export class StateStylingExtractor extends BaseGeneExtractor {
  readonly geneId: GeneId = 'state-styling';
  readonly geneName = 'State Styling';
  readonly geneDescription = 'How interactive states are styled';

  getAlleleDefinitions(): AlleleDefinition[] {
    return [
      { id: 'tailwind-variants', name: 'Tailwind State Variants', description: 'hover:, focus:, disabled:', patterns: [/className\s*=\s*["'`][^"'`]*\bhover:/, /className\s*=\s*["'`][^"'`]*\bfocus:/] },
      { id: 'css-pseudo', name: 'CSS Pseudo Classes', description: ':hover, :focus', patterns: [/:hover\s*\{/, /:focus\s*\{/] },
      { id: 'state-props', name: 'State Props', description: 'isHovered, isFocused', patterns: [/isHovered\s*[?&|]/, /isFocused\s*[?&|]/] },
      { id: 'data-state', name: 'Data State', description: 'data-state attributes', patterns: [/data-state\s*=/] },
    ];
  }

  extractFromFile(fp: string, c: string, _imp: string[]): FileExtractionResult {
    const det: DetectedAllele[] = [];
    for (const d of this.getAlleleDefinitions()) {
      for (const p of d.patterns) { const m = p.exec(c); if (m) { const ctx = this.extractContext(c, m.index); det.push({ alleleId: d.id, line: ctx.line, code: ctx.code, confidence: 0.8, context: ctx.context }); break; } }
    }
    return { file: fp, detectedAlleles: det, isComponent: this.isComponentFile(fp, c) };
  }
}

File: drift v2/crates/drift/drift-analysis/src/structural/dna/extractors/state_styling.rs
//! Frontend gene: state-styling — how component state affects styling.

use crate::structural::dna::extractor::GeneExtractor;
use crate::structural::dna::types::*;
use super::variant_handling::extract_with_definitions;

pub struct StateStylingExtractor;

impl GeneExtractor for StateStylingExtractor {
    fn gene_id(&self) -> GeneId { GeneId::StateStyling }

    fn allele_definitions(&self) -> Vec<AlleleDefinition> {
        vec![
            AlleleDefinition {
                id: "data-attributes".into(), name: "Data Attributes".into(),
                description: "Uses data-* attributes for state-based styling".into(),
                patterns: vec![r#"data-(?:state|active|disabled|selected|open|closed)"#.into()],
                keywords: vec!["data-state".into()],
                import_patterns: vec![], priority: 10,
            },
            AlleleDefinition {
                id: "aria-states".into(), name: "ARIA States".into(),
                description: "Uses aria-* attributes for state-based styling".into(),
                patterns: vec![r"aria-(?:expanded|selected|checked|pressed|disabled|hidden)".into()],
                keywords: vec!["aria-".into()],
                import_patterns: vec![], priority: 9,
            },
            AlleleDefinition {
                id: "pseudo-classes".into(), name: "CSS Pseudo-classes".into(),
                description: "Uses CSS pseudo-classes (:hover, :focus, :active)".into(),
                patterns: vec![r":(?:hover|focus|active|disabled|checked|focus-visible)".into()],
                keywords: vec![":hover".into(), ":focus".into()],
                import_patterns: vec![], priority: 6,
            },
            AlleleDefinition {
                id: "conditional-classes".into(), name: "Conditional CSS Classes".into(),
                description: "Uses conditional class application for state styling".into(),
                patterns: vec![r"is(?:Active|Open|Disabled|Selected|Loading)".into()],
                keywords: vec!["isActive".into(), "isOpen".into()],
                import_patterns: vec![], priority: 5,
            },
        ]
    }

    fn extract_from_file(&self, content: &str, file_path: &str) -> FileExtractionResult {
        extract_with_definitions(content, file_path, &self.allele_definitions())
    }
}
